agent/src/main/java/com/code_intelligence/jazzer/utils/InstrumentationResources.kt

```diff
@@ -0,0 +1,50 @@
+// Copyright 2021 Code Intelligence GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.code_intelligence.jazzer.utils
+
+import io.github.classgraph.ClassGraph
+import io.github.classgraph.ScanResult
+import kotlin.time.ExperimentalTime
+import kotlin.time.measureTimedValue
+
+@OptIn(ExperimentalTime::class)
+object InstrumentationResources {
+
+    // Class graph scan result to use in class lookups for hooks.
+    // This scan result includes class info and system jars and modules,
+    // no need for inter class dependencies (quite expensive).
+    // ScanResult should be closed manually, otherwise it's handled in a
+    // finalizer. To not introduce complicated state management this is done
+    // by the agent after instrumentation is finished.
+    var hookClassGraphResult: ScanResult?
```

Comment:
Given that this is only used from `Hook`, could we move into the companion object of that class? Since it is more narrow in scope then, we could close it right after the last call to `loadHooks` in `Agent`, potentially allowing memory to be freed sooner.

Reply:
I moved it to `Hook` and it looks ok in `Agent`. On the other hand I would like to reduce responsibilities of `Agent`.

Reply:
As discussed, I refactored the hook loading to only call it from one place within `Agent`, so that `ScanResult` can be closed properly. This turned out to be a bigger refactoring.